File: packages/php/src/Types/Metadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Document metadata.
 *
 * All fields are optional and depend on the file format, extraction configuration,
 * and postprocessors enabled.
 *
 * @property-read string|null $language Document language (ISO 639-1 code)
 * @property-read string|null $subject Document subject
 * @property-read string|null $formatType Format discriminator ("pdf", "excel", "email", etc.)
 * @property-read string|null $title Document title
 * @property-read array<string>|null $authors Document authors
 * @property-read array<string>|null $keywords Document keywords
 * @property-read string|null $createdAt Creation date (ISO 8601)
 * @property-read string|null $modifiedAt Modification date (ISO 8601)
 * @property-read string|null $createdBy Creator/application name
 * @property-read string|null $modifiedBy Modifier name
 * @property-read int|null $pageCount Number of pages
 * @property-read int|null $sheetCount Number of sheets (for spreadsheets)
 * @property-read string|null $format Image format (e.g., "PNG", "JPEG") for image documents
 * @property-read string|null $category Document category
 * @property-read array<string>|null $tags Document tags
 * @property-read string|null $documentVersion Document version
 * @property-read string|null $abstractText Document abstract text
 * @property-read string|null $outputFormat Output format used for extraction
 * @property-read array<string, mixed> $custom Additional custom metadata from postprocessors
 */
readonly class Metadata
{
    /**
     * @param array<string>|null $authors
     * @param array<string>|null $keywords
     * @param array<string>|null $tags
     * @param array<string, mixed> $custom
     */
    public function __construct(
        public ?string $language = null,
        public ?string $subject = null,
        public ?string $formatType = null,
        public ?string $title = null,
        public ?array $authors = null,
        public ?array $keywords = null,
        public ?string $createdAt = null,
        public ?string $modifiedAt = null,
        public ?string $createdBy = null,
        public ?string $modifiedBy = null,
        public ?int $pageCount = null,
        public ?int $sheetCount = null,
        public ?string $format = null,
        public ?string $category = null,
        public ?array $tags = null,
        public ?string $documentVersion = null,
        public ?string $abstractText = null,
        public ?string $outputFormat = null,
        public array $custom = [],
    ) {
    }

    /**
     * Create Metadata from array returned by extension.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $knownFields = [
            'language',
            'subject',
            'format_type',
            'title',
            'authors',
            'keywords',
            'created_at',
            'modified_at',
            'created_by',
            'modified_by',
            'page_count',
            'pageCount',
            'sheet_count',
            'format',
            'category',
            'tags',
            'document_version',
            'abstract_text',
            'output_format',
        ];

        /** @var string|null $language */
        $language = $data['language'] ?? null;

        /** @var string|null $subject */
        $subject = $data['subject'] ?? null;

        /** @var string|null $formatType */
        $formatType = $data['format_type'] ?? null;

        /** @var string|null $title */
        $title = $data['title'] ?? null;

        /** @var array<string>|null $authors */
        $authors = $data['authors'] ?? null;

        /** @var array<string>|null $keywords */
        $keywords = $data['keywords'] ?? null;

        /** @var string|null $createdAt */
        $createdAt = $data['created_at'] ?? null;

        /** @var string|null $modifiedAt */
        $modifiedAt = $data['modified_at'] ?? null;

        /** @var string|null $createdBy */
        $createdBy = $data['created_by'] ?? null;

        /** @var string|null $modifiedBy */
        $modifiedBy = $data['modified_by'] ?? null;

        /** @var int|null $pageCount */
        $pageCount = $data['page_count'] ?? $data['pageCount'] ?? null;

        /** @var int|null $sheetCount */
        $sheetCount = $data['sheet_count'] ?? null;

        /** @var string|null $format */
        $format = $data['format'] ?? null;

        /** @var string|null $category */
        $category = $data['category'] ?? null;

        /** @var array<string>|null $tags */
        $tags = $data['tags'] ?? null;

        /** @var string|null $documentVersion */
        $documentVersion = $data['document_version'] ?? null;

        /** @var string|null $abstractText */
        $abstractText = $data['abstract_text'] ?? null;

        /** @var string|null $outputFormat */
        $outputFormat = $data['output_format'] ?? null;

        $custom = [];
        foreach ($data as $key => $value) {
            if (!in_array($key, $knownFields, true)) {
                $custom[$key] = $value;
            }
        }

        return new self(
            language: $language,
            subject: $subject,
            formatType: $formatType,
            title: $title,
            authors: $authors,
            keywords: $keywords,
            createdAt: $createdAt,
            modifiedAt: $modifiedAt,
            createdBy: $createdBy,
            modifiedBy: $modifiedBy,
            pageCount: $pageCount,
            sheetCount: $sheetCount,
            format: $format,
            category: $category,
            tags: $tags,
            documentVersion: $documentVersion,
            abstractText: $abstractText,
            outputFormat: $outputFormat,
            custom: $custom,
        );
    }

    /**
     * Get a custom metadata field.
     *
     * @param string $key
     * @return mixed
     */
    public function getCustom(string $key): mixed
    {
        return $this->custom[$key] ?? null;
    }

    /**
     * Check if a custom metadata field exists.
     */
    public function hasCustom(string $key): bool
    {
        return isset($this->custom[$key]);
    }
}

File: packages/java/src/main/java/dev/kreuzberg/Metadata.java
package dev.kreuzberg;

import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Metadata extracted from a document.
 *
 * <p>
 * Contains common fields applicable to all formats (title, subject, authors,
 * keywords, language, timestamps, creators) extracted from document properties,
 * plus format-specific metadata and additional custom fields from
 * postprocessors.
 *
 * @since 0.8.0
 */
public final class Metadata {
	private final Optional<String> title;
	private final Optional<String> subject;
	private final Optional<List<String>> authors;
	private final Optional<List<String>> keywords;
	private final Optional<String> language;
	private final Optional<String> createdAt;
	private final Optional<String> modifiedAt;
	private final Optional<String> createdBy;
	private final Optional<String> modifiedBy;
	private final Optional<PageStructure> pages;
	private final Optional<Map<String, Object>> imagePreprocessing;
	private final Optional<Map<String, Object>> jsonSchema;
	private final Optional<Map<String, Object>> error;
	private final Optional<String> category;
	private final Optional<List<String>> tags;
	private final Optional<String> documentVersion;
	private final Optional<String> abstractText;
	private final Optional<String> outputFormat;
	private final Map<String, Object> additional;

	@JsonCreator
	public Metadata(@JsonProperty("title") Optional<String> title, @JsonProperty("subject") Optional<String> subject,
			@JsonProperty("authors") Optional<List<String>> authors,
			@JsonProperty("keywords") Optional<List<String>> keywords,
			@JsonProperty("language") Optional<String> language, @JsonProperty("created_at") Optional<String> createdAt,
			@JsonProperty("modified_at") Optional<String> modifiedAt,
			@JsonProperty("created_by") Optional<String> createdBy,
			@JsonProperty("modified_by") Optional<String> modifiedBy,
			@JsonProperty("pages") Optional<PageStructure> pages,
			@JsonProperty("image_preprocessing") Optional<Map<String, Object>> imagePreprocessing,
			@JsonProperty("json_schema") Optional<Map<String, Object>> jsonSchema,
			@JsonProperty("error") Optional<Map<String, Object>> error,
			@JsonProperty("category") Optional<String> category, @JsonProperty("tags") Optional<List<String>> tags,
			@JsonProperty("document_version") Optional<String> documentVersion,
			@JsonProperty("abstract_text") Optional<String> abstractText,
			@JsonProperty("output_format") Optional<String> outputFormat) {
		this.title = title != null ? title : Optional.empty();
		this.subject = subject != null ? subject : Optional.empty();
		this.authors = authors != null && authors.isPresent()
				? Optional.of(Collections.unmodifiableList(new ArrayList<>(authors.get())))
				: Optional.empty();
		this.keywords = keywords != null && keywords.isPresent()
				? Optional.of(Collections.unmodifiableList(new ArrayList<>(keywords.get())))
				: Optional.empty();
		this.language = language != null ? language : Optional.empty();
		this.createdAt = createdAt != null ? createdAt : Optional.empty();
		this.modifiedAt = modifiedAt != null ? modifiedAt : Optional.empty();
		this.createdBy = createdBy != null ? createdBy : Optional.empty();
		this.modifiedBy = modifiedBy != null ? modifiedBy : Optional.empty();
		this.pages = pages != null ? pages : Optional.empty();
		this.imagePreprocessing = imagePreprocessing != null ? imagePreprocessing : Optional.empty();
		this.jsonSchema = jsonSchema != null ? jsonSchema : Optional.empty();
		this.error = error != null ? error : Optional.empty();
		this.category = category != null ? category : Optional.empty();
		this.tags = tags != null && tags.isPresent()
				? Optional.of(Collections.unmodifiableList(new ArrayList<>(tags.get())))
				: Optional.empty();
		this.documentVersion = documentVersion != null ? documentVersion : Optional.empty();
		this.abstractText = abstractText != null ? abstractText : Optional.empty();
		this.outputFormat = outputFormat != null ? outputFormat : Optional.empty();
		this.additional = new HashMap<>();
	}

	/**
	 * Creates a new empty Metadata.
	 *
	 * @return a new empty Metadata instance
	 */
	public static Metadata empty() {
		return new Metadata(Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
				Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
				Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
				Optional.empty(), Optional.empty(), Optional.empty());
	}

	/**
	 * Used by Jackson to handle additional flattened fields from format metadata
	 * and custom postprocessor fields.
	 *
	 * @param name
	 *            the field name
	 * @param value
	 *            the field value
	 */
	@JsonAnySetter
	public void setAdditionalProperty(String name, Object value) {
		additional.put(name, value);
	}

	/**
	 * Get the document title.
	 *
	 * @return optional title
	 */
	public Optional<String> getTitle() {
		return title;
	}

	/**
	 * Get the document subject or description.
	 *
	 * @return optional subject
	 */
	public Optional<String> getSubject() {
		return subject;
	}

	/**
	 * Get the document authors.
	 *
	 * @return optional unmodifiable list of authors
	 */
	public Optional<List<String>> getAuthors() {
		return authors;
	}

	/**
	 * Get the document keywords/tags.
	 *
	 * @return optional unmodifiable list of keywords
	 */
	public Optional<List<String>> getKeywords() {
		return keywords;
	}

	/**
	 * Get the primary language code (ISO 639).
	 *
	 * @return optional language code (e.g., "en", "de")
	 */
	public Optional<String> getLanguage() {
		return language;
	}

	/**
	 * Get the creation timestamp (ISO 8601 format).
	 *
	 * @return optional creation timestamp
	 */
	public Optional<String> getCreatedAt() {
		return createdAt;
	}

	/**
	 * Get the last modification timestamp (ISO 8601 format).
	 *
	 * @return optional modification timestamp
	 */
	public Optional<String> getModifiedAt() {
		return modifiedAt;
	}

	/**
	 * Get the user who created the document.
	 *
	 * @return optional creator username
	 */
	public Optional<String> getCreatedBy() {
		return createdBy;
	}

	/**
	 * Get the user who last modified the document.
	 *
	 * @return optional modifier username
	 */
	public Optional<String> getModifiedBy() {
		return modifiedBy;
	}

	/**
	 * Get the page/slide/sheet structure with boundaries.
	 *
	 * @return optional page structure information
	 */
	public Optional<PageStructure> getPages() {
		return pages;
	}

	/**
	 * Get image preprocessing metadata (when OCR preprocessing was applied).
	 *
	 * @return optional image preprocessing metadata
	 */
	public Optional<Map<String, Object>> getImagePreprocessing() {
		return imagePreprocessing;
	}

	/**
	 * Get JSON schema (for structured data extraction).
	 *
	 * @return optional JSON schema
	 */
	public Optional<Map<String, Object>> getJsonSchema() {
		return jsonSchema;
	}

	/**
	 * Get error metadata (for batch operations).
	 *
	 * @return optional error metadata
	 */
	public Optional<Map<String, Object>> getError() {
		return error;
	}

	/**
	 * Get the document category.
	 *
	 * @return optional category string
	 * @since 4.5.0
	 */
	public Optional<String> getCategory() {
		return category;
	}

	/**
	 * Get the document tags.
	 *
	 * @return optional unmodifiable list of tags
	 * @since 4.5.0
	 */
	public Optional<List<String>> getTags() {
		return tags;
	}

	/**
	 * Get the document version.
	 *
	 * @return optional document version string
	 * @since 4.5.0
	 */
	public Optional<String> getDocumentVersion() {
		return documentVersion;
	}

	/**
	 * Get the document abstract text.
	 *
	 * @return optional abstract text
	 * @since 4.5.0
	 */
	public Optional<String> getAbstractText() {
		return abstractText;
	}

	/**
	 * Get the output format used during extraction.
	 *
	 * @return optional output format string
	 * @since 4.5.0
	 */
	public Optional<String> getOutputFormat() {
		return outputFormat;
	}

	/**
	 * Get additional custom fields from postprocessors and flattened format
	 * metadata.
	 *
	 * @return metadata map with additional fields (never null, may be empty)
	 */
	public Map<String, Object> getAdditional() {
		return Collections.unmodifiableMap(additional);
	}

	/**
	 * Check if any metadata is present.
	 *
	 * @return true if at least one field is present
	 */
	public boolean isEmpty() {
		return !title.isPresent() && !subject.isPresent() && !authors.isPresent() && !keywords.isPresent()
				&& !language.isPresent() && !createdAt.isPresent() && !modifiedAt.isPresent() && !createdBy.isPresent()
				&& !modifiedBy.isPresent() && !pages.isPresent() && !imagePreprocessing.isPresent()
				&& !jsonSchema.isPresent() && !error.isPresent() && !category.isPresent() && !tags.isPresent()
				&& !documentVersion.isPresent() && !abstractText.isPresent() && !outputFormat.isPresent()
				&& additional.isEmpty();
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (!(obj instanceof Metadata)) {
			return false;
		}
		Metadata other = (Metadata) obj;
		return Objects.equals(title, other.title) && Objects.equals(subject, other.subject)
				&& Objects.equals(authors, other.authors) && Objects.equals(keywords, other.keywords)
				&& Objects.equals(language, other.language) && Objects.equals(createdAt, other.createdAt)
				&& Objects.equals(modifiedAt, other.modifiedAt) && Objects.equals(createdBy, other.createdBy)
				&& Objects.equals(modifiedBy, other.modifiedBy) && Objects.equals(pages, other.pages)
				&& Objects.equals(imagePreprocessing, other.imagePreprocessing)
				&& Objects.equals(jsonSchema, other.jsonSchema) && Objects.equals(error, other.error)
				&& Objects.equals(category, other.category) && Objects.equals(tags, other.tags)
				&& Objects.equals(documentVersion, other.documentVersion)
				&& Objects.equals(abstractText, other.abstractText) && Objects.equals(outputFormat, other.outputFormat)
				&& Objects.equals(additional, other.additional);
	}

	@Override
	public int hashCode() {
		return Objects.hash(title, subject, authors, keywords, language, createdAt, modifiedAt, createdBy, modifiedBy,
				pages, imagePreprocessing, jsonSchema, error, category, tags, documentVersion, abstractText,
				outputFormat, additional);
	}

	@Override
	public String toString() {
		return "Metadata{" + "title=" + title + ", subject=" + subject + ", authors=" + authors + ", keywords="
				+ keywords + ", language=" + language + ", createdAt=" + createdAt + ", modifiedAt=" + modifiedAt
				+ ", createdBy=" + createdBy + ", modifiedBy=" + modifiedBy + ", pages=" + pages
				+ ", imagePreprocessing=" + imagePreprocessing + ", jsonSchema=" + jsonSchema + ", error=" + error
				+ ", category=" + category + ", tags=" + tags + ", documentVersion=" + documentVersion
				+ ", abstractText=" + abstractText + ", outputFormat=" + outputFormat + ", additional=" + additional
				+ '}';
	}
}
